google/cloud/internal/future_then_async.h

```diff
@@ -0,0 +1,89 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_FUTURE_THEN_ASYNC_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_FUTURE_THEN_ASYNC_H
+/**
+ * @file
+ *
+ * Define the `then_async` template function.
+ *
+ * This function allows for generic asynchronous continuations using
+ * google::cloud::future. In particular, it allows chaining of async
+ * operations without increasing the nesting level.
+ * 
+ * Usage of this utility would look like the below, which is a series of
+ * asynchronous operations written in an almost linear style:
+ * 
+ * future<std::string> f1 = ...
+ * future<int> f2 = then_async(std::move(f1), [](std::string value) -> future<int> {
```

Comment:
Note that `google::cloud::future<>` already unwraps:

https://github.com/googleapis/google-cloud-cpp/blob/6600f0fda6f6da0c8d248beadbc519db23310f94/google/cloud/future_generic.h#L95-L96

I think your implementation could be:

```cc
template <typename T, typename F>
auto then_async(future<T> f, F&& f) /* -> this could be ugly, look at internal::then_helper<F, T>::future_t */ {
  return f.then([g = std::forward<Functor>(f)](future<T> h) { return g(h.get()); });
}
```